Userland/Libraries/LibArchive/Lzma.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2023, Tim Schumacher <timschumi@gmx.de>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/MaybeOwned.h>
+#include <AK/NonnullOwnPtr.h>
+#include <AK/Stream.h>
+#include <LibCompress/Lzma.h>
+
+namespace Archive {
+
+// This implementation is mostly based on the LZMA specification contained in the 7-Zip SDK, which has been placed in the public domain.
+// LZMA Specification Draft (2015): https://www.7-zip.org/a/lzma-specification.7z
+
+// Described in section "lzma file format".
+struct [[gnu::packed]] LzmaHeader {
+    u32 dictionary_size() const;
+    Optional<u64> uncompressed_size() const;
+
+    ErrorOr<void> decode_model_properties(u8& literal_context_bits, u8& literal_pos_bits, u8& pos_bits) const;
+
+    ErrorOr<Compress::LzmaDecompressorOptions> as_decompressor_options() const;
+
+private:
+    u8 m_model_properties;
+    u32 m_dictionary_size;
+    u64 m_uncompressed_size;
+};
+static_assert(sizeof(LzmaHeader) == 13);
+
+class LzmaDecompressor {
+public:
+    static ErrorOr<NonnullOwnPtr<Compress::LzmaDecompressor>> create(MaybeOwned<Stream>);
```

Comment:
Archive::LzmaDecompressor::create() returns a Compress::LzmaDecompressor? This seems kind of odd, but then I didn't cross reference it with the other Decompressors that mostly just redirect to LibCompress.

Reply:
The reason why I made this separate is because LZMA technically defines an 'archive' format (which is more of a stream container, really).

It is a really puny archive format, because all it does is prefixing it with a small header that notes down the decoder settings for the contained stream, but it is distinct from the other LibCompress-supported algorithms that really only dump the raw stream into a file and call it a day.

The reason why it returns a `Compress::LzmaDecompressor` is because we read the header upon creation of the stream, configure the underlying `LibCompress` stream once, and would then just delegate all the time anyways. Returning the underlying Stream directly and not implementing the whole Stream interface for Archive::LzmaDecompressor again seemed like a better idea.

Would we rather want to have a `Compress::LzmaDecompressor::create_from_archive` (/container?) and `Compress::LzmaDecompressor::create_from_raw_stream`, where one relies on the other?

Reply:
From Idans comment about the same topic:

> The distiniction between a compression format and an archival format, is that a compression format takes a stream of bytes and compresses/decompresses it, while an archive format packs together one or more files, with the file contents possibly being compressed using a compression format.

Ok, in that case LZMA 'archives' wouldn't be an actual archive because (in comparison to an LZMA compressed stream) they only provide the settings for the singular stream that they contain.

However, looking at what's to come, are LZMA2 and XZ archive formats then? Should they also rather be in LibCompress?

Neither of them have a concept of one or multiple files, both just encode a stream of data, but this time in multiple ways. LZMA2 splits the stream in parts and either stores them LZMA-compressed or uncompressed. XZ allows compressing its data stream using LZMA2 and/or a bunch of of other so-called filters. None of them know what 'files' are, if a user needs that they have to TAR it up first.

Reply:
Comparing to the Gzip, which we already have implemented in serenity, it seems to match up: It's based on Deflate, it adds a short header at the start with the compression options, and is possibly used to compress tar archives.
Currently the implementation for Gzip is in LibCompress, matching my expectations at least.

The only unfortunate part about LZMA is that the name is used both for the raw stream and the stream with the header part.

I would say all 3 of LZMA, LZMA2 and XZ should end up in LibCompress.